### PR TITLE
ccache: update to 4.0 (alternative approach for chroot-cmake)

### PR DESCRIPTION
**Pull request by sgn**
Alternative for https://github.com/void-linux/void-packages/pull/25724

### CODE REVIEW

**Comment by pullmoll** [comment]
I like your approach. If we don't need more than what a `cmake bootstrap` supplies that would be very space saving.

**Review by pullmoll — APPROVED** [review]

File: srcpkgs/ccache/template
Comment (sgn) [reply srcpkgs/ccache/template:17]: I was thinking about `stage1` and `stage2`.
Either put inside `chroot_ready` or set it unconditionally. I don't really know. I haven't put much thought into that. :shrug:

Comment (pullmoll) [reply srcpkgs/ccache/template:17]: As a rule of thumb you usually do not depend on anything `chroot-...` because it is, well, in the chroot anyway.
That's why I thought that perhaps you meant
```
if [ "$CHROOT_READY" -a "$CROSS_BUILD" ]; then
        hostmakedepends+=" cmake"
fi
```
We need to test that anyway... I'll close my PR, your's is better.

Comment (sgn) [reply srcpkgs/ccache/template:17]: This `chroot-cmake` is very basic `cmake`, it lacks a lot of `cmake` functionality (but enough to build `ccache` and real `cmake`).
I think we should treat it the same with `pkgconf` and `pkg-config`. IOW, don't add it into `base-chroot` but let's package add it into `hostmakedepends`.

Comment (pullmoll) [reply srcpkgs/ccache/template:17]: Ok, I see, then this makes sense.
Let's hope that ccache doesn't become too fancy with its cmake scripts in the next updates :)

Comment (pullmoll) [reply srcpkgs/ccache/template:17]: If we don't add it to `base-chroot` then a better name then would be something like `cmake-minimal` or `cmake-bootstrap`.

Comment (sgn) [reply srcpkgs/ccache/template]: OK, I'll change it into `cmake-bootstrap`

Comment (sgn) [reply srcpkgs/ccache/template]: Looking around, I think if `ccache` don't use those fancy  `ExternalProject`, we should be fine.

The whole `curl`, `libarchive`, `rhash`, `expat` and `nghttp2` was added for fetching tarball from external resource, verify checksum and build them.

**Comment by pullmoll** [comment]
Running `./xbps-src bootstrap` now behaves odd.
It seems to enter an endless loop trying to build `ncurses`.
When I break with `^C` and redo the command it succeeds.
Not sure what kind of dependency loop (?) is causing this.

**Comment by pullmoll** [comment]
In `srcpkgs/ccache/template` add the following lines to fix cross builds:
```
if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
        makedepends+=" libatomic-devel"
fi

pre_configure() {
        if [ "$XBPS_TARGET_NO_ATOMIC8" ]; then
                vsed -i src/CMakeLists.txt \
                        -e "/add_library/a target_link_libraries(ccache_lib PRIVATE atomic)"
        fi
}
```

File: srcpkgs/cmake-bootstrap/template
Comment (q66) [inline srcpkgs/cmake-bootstrap/template:14]: bad short_desc, cmake is not a "cross make" or anything of the sort

**Comment by pullmoll** [comment]
On a 2nd try the `./xbps-src zap` and `./xbps-src bootstrap` succeeded to build stage0. No idea what was causing the loop in the 1st try.

File: srcpkgs/cmake-bootstrap/template
Comment (sgn) [reply srcpkgs/cmake-bootstrap/template]: Heh, they self-claimed "cross-platform make" and I was lazy. I'll be more lazy to just write "CMake"

**Comment by sgn** [comment]
> Running `./xbps-src bootstrap` now behaves odd.
> It seems to enter an endless loop trying to build `ncurses`.
> When I break with `^C` and redo the command it succeeds.
> Not sure what kind of dependency loop (?) is causing this.
> 
> Did you try a `bootstrap`?

Still running, my main machine is going down, so I need to `bootstrap` on real crappy machine now.

**Comment by pullmoll** [comment]
I've been testing it and already successfully built [x86_64 bootstrap stage0 + stage1, cross base-chroot for i686 and x86_64-musl, and several thousand x86_64 packages](https://repo.voidlinux.de) without issues so far. IMO this PR can be merged.

**Comment by sgn** [comment]
I'll let someone else press the merge button. I've received enough blame those days.

File: srcpkgs/cmake-bootstrap/template
Comment (ericonr) [inline srcpkgs/cmake-bootstrap/template:29]: why is this commented?

File: srcpkgs/cmake-bootstrap/template
Comment (ericonr) [inline srcpkgs/cmake-bootstrap/template:39]: shouldn't this be "full cmake" ? It's still real in that it can build packages.

File: srcpkgs/cmake-bootstrap/template
Comment (sgn) [reply srcpkgs/cmake-bootstrap/template:29]: Originally, I wanted to remove all 3rd party libraries except `libuv` since I know that `libuv` is required. I found out `jsoncpp` was required later on.

File: srcpkgs/cmake-bootstrap/template
Comment (sgn) [reply srcpkgs/cmake-bootstrap/template:39]: I used that because CMake uses `real real cmake` to talk about `cmake`, I'll change description to "this is half baked cmake"

**Comment by sgn** [comment]
Ready again, now!

**Review by pullmoll — APPROVED** [review]

**Comment by Piraty** [comment]
why all the pain?
why not remove ccache from bootstrap packages, i don't see the necessity to have it that early really.
With ccache being `bootstrap=yes` we can't make it build its manpages, as they need [asciidoc](https://github.com/ccache/ccache/blob/v4.0/doc/CMakeLists.txt#L4) and [perl](https://github.com/ccache/ccache/blob/v4.0/doc/CMakeLists.txt#L52).

**Comment by sgn** [comment]
For future reference, the discussion is on #26034

**Pull request merged**
